inc/saiswitch.h

```diff
@@ -325,9 +325,151 @@ typedef enum _sai_switch_failover_config_mode_t
 
 } sai_switch_failover_config_mode_t;
 
+/**
+ * @brief Defines tunnel type
+ */
+typedef enum _sai_tunnel_type_t
+{
+    SAI_TUNNEL_TYPE_IPINIP,
+
+    SAI_TUNNEL_TYPE_IPINIP_GRE,
+
+    SAI_TUNNEL_TYPE_VXLAN,
+
+    SAI_TUNNEL_TYPE_MPLS,
+
+} sai_tunnel_type_t;
+
+/**
+ * @brief Defines tunnel encap ECN mode
+ */
+typedef enum _sai_tunnel_encap_ecn_mode_t
+{
+    /**
+     * @brief Normal mode behavior defined in RFC 6040
+     * section 4.1 copy from inner
+     */
+    SAI_TUNNEL_ENCAP_ECN_MODE_STANDARD,
+
+    /**
+     * @brief User defined behavior.
+     */
+    SAI_TUNNEL_ENCAP_ECN_MODE_USER_DEFINED
+
+} sai_tunnel_encap_ecn_mode_t;
+
+/**
+ * @brief Defines tunnel decap ECN mode
+ */
+typedef enum _sai_tunnel_decap_ecn_mode_t
+{
+    /**
+     * @brief Behavior defined in RFC 6040 section 4.2
+     */
+    SAI_TUNNEL_DECAP_ECN_MODE_STANDARD,
+
+    /**
+     * @brief Copy from outer ECN
+     */
+    SAI_TUNNEL_DECAP_ECN_MODE_COPY_FROM_OUTER,
+
+    /**
+     * @brief User defined behavior
+     */
+    SAI_TUNNEL_DECAP_ECN_MODE_USER_DEFINED
+
+} sai_tunnel_decap_ecn_mode_t;
+
+/**
+ * @brief Defines tunnel attributes at switch level.
+ * SAI_OBJECT_TYPE_SWITCH_TUNNEL object provides
+ * per tunnel type global configuration.
+ * SAI_OBJECT_TYPE_TUNNEL object configuration
+ * overrides the switch scoped global configuration.
+ */
+typedef enum _sai_switch_tunnel_attr_t
+{
+    /**
+     * @brief Start of attributes
+     */
+    SAI_SWITCH_TUNNEL_ATTR_START,
+
+    /**
+     * @brief Tunnel type key
+     *
+     * @type sai_tunnel_type_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY | KEY
+     * @isresourcetype true
+     */
+    SAI_SWITCH_TUNNEL_ATTR_TUNNEL_TYPE = SAI_SWITCH_TUNNEL_ATTR_START,
+
+    /**
+     * @brief Packet action when a packet ingress and gets routed back to same tunnel
+     *
+     * @type sai_packet_action_t
+     * @flags CREATE_AND_SET
+     * @default SAI_PACKET_ACTION_FORWARD
+     */
+    SAI_SWITCH_TUNNEL_ATTR_LOOPBACK_PACKET_ACTION,
+
+    /* Tunnel encap attributes */
+
+    /**
+     * @brief Tunnel encap ECN mode
+     *
+     * @type sai_tunnel_encap_ecn_mode_t
+     * @flags CREATE_ONLY
+     * @default SAI_TUNNEL_ENCAP_ECN_MODE_STANDARD
+     */
+    SAI_SWITCH_TUNNEL_ATTR_TUNNEL_ENCAP_ECN_MODE,
+
+    /**
+     * @brief Tunnel encap mappers
+     *
+     * @type sai_object_list_t
+     * @flags CREATE_ONLY
+     * @objects SAI_OBJECT_TYPE_TUNNEL_MAP
+     * @default empty
+     */
+    SAI_SWITCH_TUNNEL_ATTR_ENCAP_MAPPERS,
```

Comment:
This will allow for any kind of mapper to be defined at a switch level (VLAN2VNI, VRF2VNI etc.). I assume it was not intended and we need to mention it at least in the comment, or better to allow only for specific types of mappers that are switch scoped.

Reply:
Added comments to intent that switch scoped mappers are for ECN mappers only
https://github.com/opencomputeproject/SAI/pull/1173/commits/ff7b05fb5c25a5d609fa5c0039d30f0e349596a3